### DISCUSSION TITLE
How to implement Expand operator of Spark in Velox?

### DISCUSSION THREAD

**Discussion by JkSelf** [facebookincubator/velox#5958] [Ideas]
Spark utilizes the Expand operator to handle ROLLUP functions, yet it differs from [GroupID](https://facebookincubator.github.io/velox/develop/operators.html#groupidnode). The Expand operator focuses on `projections`, including `aggregation `expressions, `grouping `expressions, and `spark_grouping_id`. Its purpose is to demonstrate how a single row can be expanded into multiple rows. However, unlike GroupID, the Expand operator in Spark cannot explicitly distinguish between the `aggregation `key and the `grouping `key based on `projections `alone.

Take the lineitem table in TPCH as an example. And the query is `select sum(l_suppkey) from lineitem where group by ROLLUP(l_orderkey, l_partkey)`.  The physical plan in spark is below.

```
== Physical Plan ==\n
AdaptiveSparkPlan isFinalPlan=false\n
+- HashAggregate(keys=[l_orderkey#42L, l_partkey#43L, spark_grouping_id#41L], functions=[sum(l_suppkey#2L)])\n
   +- Exchange hashpartitioning(l_orderkey#42L, l_partkey#43L, spark_grouping_id#41L, 200), ENSURE_REQUIREMENTS, [plan_id=26]\n
      +- HashAggregate(keys=[l_orderkey#42L, l_partkey#43L, spark_grouping_id#41L], functions=[partial_sum(l_suppkey#2L)])\n
         +- Expand [[l_suppkey#2L, l_orderkey#0L, l_partkey#1L, 0], [l_suppkey#2L, l_orderkey#0L, null, 1], [l_suppkey#2L, null, null, 3]], [l_suppkey#2L, l_orderkey#42L, l_partkey#43L, spark_grouping_id#41L]\n
            +- BatchScan[l_orderkey#0L, l_partkey#1L, l_suppkey#2L] ParquetScan DataFilters: [], Format: parquet, Location: InMemoryFileIndex(1 paths)[file:/home/sparkuser/jk/projects/gluten-3.2/gluten-core/src/test/resou..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<l_orderkey:bigint,l_partkey:bigint,l_suppkey:bigint>, PushedFilters: [] RuntimeFilters: []\n\n|

```
The input and output for Expand and HashAggregate operators are illustrated below. It is hard to distinguish the aggregation key and grouping key in Expand operator. So we added a new Expand operator in Velox to align with Spark. See https://github.com/facebookincubator/velox/pull/5403
```
(10) Expand
Input [3]: [l_orderkey#0L, l_partkey#1L, l_suppkey#2L]
Arguments: [[l_suppkey#2L, l_orderkey#0L, l_partkey#1L, 0], [l_suppkey#2L, l_orderkey#0L, null, 1], [l_suppkey#2L, null, null, 3]], [l_suppkey#2L, l_orderkey#59L, l_partkey#60L, spark_grouping_id#58L]

(11) HashAggregate
Input [4]: [l_suppkey#2L, l_orderkey#59L, l_partkey#60L, spark_grouping_id#58L]
Keys [3]: [l_orderkey#59L, l_partkey#60L, spark_grouping_id#58L]
Functions [1]: [partial_sum(l_suppkey#2L)]
Aggregate Attributes [1]: [sum#68L]
Results [4]: [l_orderkey#59L, l_partkey#60L, spark_grouping_id#58L, sum#69L]
```

**Comment by JkSelf** [comment]
@mbasmanova

**Comment by mbasmanova** [comment]
@JkSelf Thank you for the examples. It is very helpful. I'm hesitant to introduce this operator to Velox since it feels legacy. GroupId is more elegant, easier to reason about and less verbose solution for ROLLUP queries. It feels like it should be possible to identify a pattern of Expand followed by Aggregation and convert it to GroupId followed by Aggregation. Is this something you would consider implementing in Gluten?

**Reply by JkSelf** [reply]
Our initial approach was to use the `GroupID `operator to implement the `Expand `behavior. We assumed that projections were in the form of `(agg exprs+ grouping exprs+ gid)`, and then get the `agg exprs` from `Aggregate `to split the `projections`. But the `projections ` has the following two situations besides the above case:

```
agg exprs + group by exprs + gid + _gen_grouping_pos. --> The last column is for handling duplicate grouping sets.
group by exprs + gid + agg exprs. --> gid is calculated by different way from above two cases. it's assigned with the sequence number of project set.
```
If these three situations are supported, the code on the Gluten side will be particularly complicated and difficult to maintain.

And now there are two backends behind gluten: velox and clickhouse. For the clickhouse backend, they do not need to distinguish the `projections `of Expand.

**Reply by mbasmanova** [reply]
@JkSelf Thank you for further explanation. Some follow-up questions.

> agg exprs + group by exprs + gid + _gen_grouping_pos. --> The last column is for handling duplicate grouping sets.

Can you explain this use case a bit more? What kind of query is using this?

**Reply by JkSelf** [reply]
The query is `select count(l_suppkey) from lineitem group by grouping sets((l_suppkey), (l_suppkey), (l_orderkey))`. And the spark plan is as below.
```
== Physical Plan ==\n
 AdaptiveSparkPlan isFinalPlan=false\n
 +- HashAggregate(keys=[l_suppkey#43L, l_orderkey#44L, spark_grouping_id#41L, _gen_grouping_pos#42], functions=[count(l_suppkey#2L)])\n   
    +- Exchange hashpartitioning(l_suppkey#43L, l_orderkey#44L, spark_grouping_id#41L, _gen_grouping_pos#42, 200), ENSURE_REQUIREMENTS, [plan_id=23]\n 
       +- HashAggregate(keys=[l_suppkey#43L, l_orderkey#44L, spark_grouping_id#41L, _gen_grouping_pos#42], functions=[partial_count(l_suppkey#2L)])\n
          +- Expand [[l_suppkey#2L, l_suppkey#2L, null, 1, 0], [l_suppkey#2L, l_suppkey#2L, null, 1, 1], [l_suppkey#2L, null, l_orderkey#0L, 2, 2]], [l_suppkey#2L, l_suppkey#43L, l_orderkey#44L, spark_grouping_id#41L, _gen_grouping_pos#42]\n
             +- Project [l_suppkey#2L, l_suppkey#2L, l_orderkey#0L]\n  
                +- BatchScan[l_orderkey#0L, l_suppkey#2L] ParquetScan DataFilters: [], Format: parquet, Location: InMemoryFileIndex(1 paths)[file:/home/sparkuser/jk/projects/gluten-3.2/gluten-core/src/test/resou..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<l_orderkey:bigint,l_suppkey:bigint>, PushedFilters: [] RuntimeFilters: []\n\n|

```

**Reply by mbasmanova** [reply]
@JkSelf Got it. Thank you for the example. I wonder why would Spark plan this query like so. it seems wasteful to compute same aggregation twice. Wouldn't it be more efficient to aggregate only by unique grouping sets, then use projection to duplicate some of the results?

I believe I now understand how Expand works and how it is used in Spark. BTW, are these any other use cases that I need to be aware of?

It very much looks like Expand can be implemented as Project followed by Unnest. Project operator would produce arrays (one array per column). Something like this.

```
select sum(l_suppkey) from lineitem where group by ROLLUP(l_orderkey, l_partkey)

select t.* from tpch.tiny.lineitem, unnest(array[suppkey, suppkey, suppkey], array[orderkey, orderkey, null], array[suppkey, null, null], array[0, 1, 2]) as t(s, o, p, gid)
where orderkey = 1;

         - Unnest[replicate=, unnest=expr:array(bigint), expr_3:array(bigint), expr_4:array(bigint), expr_5:array(integer)] => [field:bigint, field_0:bigint, field_1:bigint, field_2:integer]       >
             - ScanFilterProject[table = TableHandle {connectorId='tpch', connectorHandle='lineitem:sf0.01', layout='Optional[lineitem:sf0.01]'}, filterPredicate = (orderkey) = (BIGINT'1'), project>
                     Estimates: {source: CostBasedSourceInfo, rows: 60175 (1.84MB), cpu: 1083150.00, memory: 0.00, network: 0.00}/{source: CostBasedSourceInfo, rows: 4 (128B), cpu: 2166300.00, memo>
                     expr := ARRAY(suppkey, suppkey, suppkey) (1:26)                                                                                                                                 >
                     expr_3 := ARRAY(orderkey, orderkey, null) (1:26)                                                                                                                                >
                     expr_4 := ARRAY(suppkey, null, null) (1:26)                                                                                                                                     >
                     expr_5 := [Block: position count: 3; size: 84 bytes]                                                                                                                            >
                     suppkey := tpch:suppkey (1:25)                                                                                                                                                  >
                     orderkey := tpch:orderkey (1:25)
```

Is there any particular reason not to implement Expand as Project + Unnest?

**Reply by JkSelf** [reply]
@mbasmanova  Thank you very much for your detailed explanation. It looks like we can implement the Expand operator by adding a Project + unnest. Then we may no need [5403 ](https://github.com/facebookincubator/velox/pull/5403). And I will try to implement Expand in Gluten directly. 

> I wonder why would Spark plan this query like so. it seems wasteful to compute same aggregation twice. Wouldn't it be more efficient to aggregate only by unique grouping sets, then use projection to duplicate some of the results?

Yes. Thanks for your suggestions. I will have a try in Gluten.


> I believe I now understand how Expand works and how it is used in Spark. BTW, are these any other use cases that I need to be aware of?

There are currently no other cases that require special consideration.

**Reply by JkSelf** [reply]
@mbasmanova I optimized the Expand operator in gluten by implementing it using Project and Unnest. However, this change resulted in a performance drop from 278s to 329s for Q67 in 2TB TPC-DS. The execution time increased by a factor of 1.18x. The primary reason for this degradation is that the data is processed on both sides of Project and Expand, leading to a doubling of the output bytes from the Expand operator.

**Reply by mbasmanova** [reply]
> The primary reason for this degradation is that the data is processed on both sides of Project and Expand, leading to a doubling of the output bytes from the Expand operator.
 
 @JkSelf Thank you for experimenting. Would you clarify what it means exactly that "data is processed on both sides of Project and Expand" and why Expand operator emits 2x more bytes?

**Reply by JkSelf** [reply]
@mbasmanova I apologize for my mistake. The data processing occurs on both sides of `Project and Unnest` operators, not `Project and Expand`. Presently, the spark Expand operator in Gluten counts the total number of output bytes by adding the metrics from `Project and Unnest` operators, resulting in a doubling of the output bytes. Currently, it appears that the 
`Project and Unnest` approach can achieve Expand functionality. However, due to the need for both the `Project and Unnest` operators to iterate over the input data, it results in a performance decrease.

**Reply by mbasmanova** [reply]
@JkSelf 

>  the spark Expand operator in Gluten counts the total number of output bytes by adding the metrics from Project and Unnest operators, resulting in a doubling of the output bytes. 

Sounds like an accounting bug, no? There is no actual doubling of the output, right?

> due to the need for both the Project and Unnest operators to iterate over the input data, it results in a performance decrease.

Would it be possible to put together a small benchmark to illustrate this? What do you see in the profile? What is taking time? How much perf difference there is between Project + Unnest and Expand?

**Reply by JkSelf** [reply]
> Sounds like an accounting bug, no? There is no actual doubling of the output, right?
Yes. 

> Would it be possible to put together a small benchmark to illustrate this? What do you see in the profile? What is taking time? How much perf difference there is between Project + Unnest and Expand?

Sure. I will add a simple benchmark to illustrate this issue later.

**Reply by JkSelf** [reply]
@mbasmanova I have added a simple benchmark(`TableScan + Project  + Unnest`  vs `TableScan + Expand`) to compare the performance between ProjectAndUnnest and Expand operations on the TPCH lineitem table in Velox. The size of the Parquet file used for testing is approximately 1.1GB. You can find the code for this benchmark in the latest commit [here](https://github.com/JkSelf/velox/tree/add-expand-op).

The results of the benchmark are as follows: 
```
============================================================================
[...]lox/benchmarks/tpch/TpchBenchmark.cpp     relative  time/iter   iters/s
============================================================================
projectAndUnnest                                            10.26s    97.44m
Expand                                                       2.20s   455.35m

```

The Expand operation completes in just 2.20 seconds, whereas the ProjectAndUnnest operation takes around 10.26 seconds.
This benchmark provides insights into the significant performance improvement achieved by using the Expand operation over ProjectAndUnnest on the given dataset.

**Reply by mbasmanova** [reply]
@JkSelf This is interesting. Have you looked at the profile to see where the slowness come from? Let's make sure this is a not some silly bug. Assuming it is not and we want to proceed with introducing Expand operator, let's do the following:

1- Update PR description to explain that functionally Expand operator can be implemented using Unnest, but performance is better if this logic is implemented as a separate operator. Make it clear that we introduce Expand for performance reasons.
2- Update documentation to clearly describe the Expand operator. It is super hard to understand, hence, let's include examples of input and output and describe cases when Spark uses this operator. Let's make it possible for someone not familiar with the operator to understand how it works without asking lots of questions.
3- Remove [GLUTEN] prefix from the PR title. 
4- Add separate test file for Expand operator specifically (similar to UnnestTest.cpp). Make sure to cover all cases.
5- Make sure to test serde for the new plan node.
6- Make sure PlanNodeBuilder::expand is document like all other methods in this class.

**Reply by JkSelf** [reply]
@mbasmanova 
I utilized `perf `to collect hotspot data, and the analysis reveals the performance degradation originating from the `copyValuesAndNulls ` method while constructing the `ArrayConstructor` [here](https://github.com/facebookincubator/velox/blob/f943531fd9cab66c4d9abf6e301ff5a4264b6283/velox/functions/prestosql/ArrayConstructor.cpp#L29).
![image](https://github.com/facebookincubator/velox/assets/11972570/8efb2f50-6edd-4a65-87ea-3d08ef5c309e)

**Reply by mbasmanova** [reply]
@JkSelf Thank you for profiling. Sounds like we are spending a lot of time in projection creating arrays. Is this right? Would you share the output of printPlanWithStats that should illustrate the breakdown of CPU time between project and unnest operators. 

Have you looked into why copyValuesAndNulls is so slow? Does it copy values one row at a time?

**Reply by JkSelf** [reply]
@mbasmanova 
Here is the printPlanWithStats  of Expand:
```
--Expand[[l_orderkey, l_partkey, l_suppkey, 0], [l_orderkey, null, l_suppkey, 1], [null, null, l_suppkey, 2]] -> l_orderkey:BIGINT, l_partkey:BIGINT, l_suppkey:BIGINT, group_id_0:BIGINT
   Output: 80357184 rows (1.47GB, 8064 batches), Cpu time: 92.50ms, Blocked wall time: 0ns, Peak memory: 256B, Memory allocations: 8064, Threads: 4
  -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
     Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.17s, Blocked wall time: 0ns, Peak memory: 313.31MB, Memory allocations: 14270, Threads: 4, Splits: 10
```

Here is the printPlanWithStats  of Project and Unnest:
```
-- Unnest[c0, c1, c2] -> c0_e:BIGINT, c1_e:BIGINT, c2_e:BIGINT
   Output: 80357184 rows (1.98GB, 2688 batches), Cpu time: 2.73s, Blocked wall time: 0ns, Peak memory: 356.00KB, Memory allocations: 21504, Threads: 4
  -- Project[expressions: (c0:ARRAY<BIGINT>, array_constructor(ROW["l_suppkey"],ROW["l_suppkey"],ROW["l_suppkey"])), (c1:ARRAY<BIGINT>, array_constructor(ROW["l_orderkey"],ROW["l_orderkey"],null)), (c2:ARRAY<BIGINT>, array_constructor(ROW["l_partkey"],null,null))] -> c0:ARRAY<BIGINT>, c1:ARRAY<BIGINT>, c2:ARRAY<BIGINT>
     Output: 26785728 rows (2.72GB, 2688 batches), Cpu time: 21.04s, Blocked wall time: 0ns, Peak memory: 1.04MB, Memory allocations: 40332, Threads: 4
    -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
       Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.07s, Blocked wall time: 0ns, Peak memory: 308.52MB, Memory allocations: 16519, Threads: 4, Splits: 10

```

The CPU time of Project is 21.04s. And the performance hotspot appears to be within the Project operation during array data creation. The ArrayConstructor utilizes the copyRange method to copy data [here](https://github.com/JkSelf/velox/blob/166cd0a6e118ba26e12feb3d734a906cbb3f0d19/velox/functions/prestosql/ArrayConstructor.cpp#L69). However, it seems that the range count is consistently 1, which could be the cause of the performance degradation. Next, should we continue to optimize the performance of Project and unnest operations or introduce a new expand operator?

**Reply by mbasmanova** [reply]
@JkSelf Thank you for sharing more performance info. This is very useful.

> However, it seems that the range count is consistently 1, which could be the cause of the performance degradation. 

Any chance you can debug some more to find out why range count is 1?

> Next, should we continue to optimize the performance of Project and unnest operations or introduce a new expand operator?

We may need to do both. I suggest to continue investigating the performance of array_constructor. It would be nice to get to the bottom of this and fix. At the same time, let's also proceed with adding Expand operator.

Thanks.

**Reply by JkSelf** [reply]
@mbasmanova Sure. I will continue investigating the performance of array_constructor later.

**Reply by JkSelf** [reply]
@mbasmanova The range count is set to 1 [here](https://github.com/facebookincubator/velox/blob/aa1728f95c84307679d9d9d74aa583a96b81d189/velox/functions/prestosql/ArrayConstructor.cpp#L65) by you. And the PR is https://github.com/facebookincubator/velox/pull/5985/files. Do you have any input?

**Reply by mbasmanova** [reply]
@JkSelf  Here, we need to copy values from input vector into target 'elements' vector with "gaps". E.g. rows 0, 1, 2, 3 need to go into rows 0, 5, 9, 13. Hence, it is correct that we create one range per value, but FlatVector::copyRanges can be optimized to copy values in bulk. 

Here is current implementation.

FlatVector.h

```
  void copyRanges(
      const BaseVector* source,
      const folly::Range<const BaseVector::CopyRange*>& ranges) override {
    for (auto& range : ranges) {
      copy(source, range.targetIndex, range.sourceIndex, range.count);
    }
  }
```

One option would be to construct a toSourceRow mapping for all ranges and call copyValuesAndNulls. Would you try this to see if it helps?

```
void copyValuesAndNulls(
      const BaseVector* source,
      const SelectivityVector& rows,
      const vector_size_t* toSourceRow);
```

**Reply by JkSelf** [reply]
@mbasmanova When attempting to construct the `toSourceRow `method, I encountered a situation where we need to splice numArgs into a vector before invoking the copy function. However, this approach introduces additional copy operations. Is this the expected behavior? Some code looks like this.

     `auto indices = AlignedBuffer::allocate<vector_size_t>(
          rows.end() * numArgs, context.pool());
      auto toSourceRow = indices->asMutable<vector_size_t>();
      // Construct toSourceRow
      for (int i = 0; i < rows.end() * numArgs; i++) {
        if (i % numArgs == 0) {
	    toSourceRow[i] = i / numArgs;
        } else {
           toSourceRow[i] = (i % numArgs) * rows.end() + i / numArgs;
        }
      }
      // Splice the numArgs into input vector.
      VectorPtr input = args[0];
      for (int i = 1; i < numArgs; i++) {
        input->append(args[i].get());
      }

      SelectivityVector vectors(numArgs * rows.countSelected());
      vectors.setAll();
      elementsResult->copy(input, vectors, toSourceRow);
      vectors.clearAll();`

**Reply by mbasmanova** [reply]
@JkSelf This is not quite what I had in mind.

Let's say we construct array from 2 values: array(a, b). If a is a vector of [a1, a2, a3] and b a vector of [b1, b2, b3], then we want to create elements vector [a1, b1, a2, b2, a3, b3].

To do that, we allocate a vector of size 6:[_, _, _, _, _, _, ].

Then copy values from a: [a1, _, a2, _, a3, _].

Then copy values from b: [a1, b1, a2, b2, a3, b3].

To copy values from a, we call copyValuesAndNulls with rows = [0, 2, 4] and toSourceRow = [0, 0, 1, 0, 2, 0].

To copy values from b, we call copyValuesAndNulls with rows = [1, 3, 5] and toSourceRow = [0, 0, 0, 1, 0, 2].

Hope this helps.

**Reply by mbasmanova** [reply]
@JkSelf Here is what this might look like: https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:opt-array-constructor?expand=1

**Reply by JkSelf** [reply]
@mbasmanova Thanks for your help. This method facilitates a reduction in project time from 21s to 10s. However, the performance is still slower compared to the Expand operation.

Here is the printPlanWithStats of Expand:
```
-- Expand[[l_orderkey, l_partkey, l_suppkey, 0], [l_orderkey, null, l_suppkey, 1], [null, null, l_suppkey, 2]] -> l_orderkey:BIGINT, l_partkey:BIGINT, l_suppkey:BIGINT, group_id_0:BIGINT
   Output: 80357184 rows (1.47GB, 8064 batches), Cpu time: 101.74ms, Blocked wall time: 0ns, Peak memory: 256B, Memory allocations: 8064, Threads: 4
  -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
     Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.32s, Blocked wall time: 0ns, Peak memory: 308.52MB, Memory allocations: 15388, Threads: 4, Splits: 10
```

Here is the printPlanWithStats of Project and Unnest after copyValuesAndNulls  optimization:
```
-- Unnest[c0, c1, c2] -> c0_e:BIGINT, c1_e:BIGINT, c2_e:BIGINT
   Output: 80357184 rows (1.98GB, 2688 batches), Cpu time: 2.82s, Blocked wall time: 0ns, Peak memory: 356.00KB, Memory allocations: 21504, Threads: 4
  -- Project[expressions: (c0:ARRAY<BIGINT>, array_constructor(ROW["l_suppkey"],ROW["l_suppkey"],ROW["l_suppkey"])), (c1:ARRAY<BIGINT>, array_constructor(ROW["l_orderkey"],ROW["l_orderkey"],null)), (c2:ARRAY<BIGINT>, array_constructor(ROW["l_partkey"],null,null))] -> c0:ARRAY<BIGINT>, c1:ARRAY<BIGINT>, c2:ARRAY<BIGINT>
     Output: 26785728 rows (2.72GB, 2688 batches), Cpu time: 10.29s, Blocked wall time: 0ns, Peak memory: 1.04MB, Memory allocations: 40332, Threads: 4
    -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
       Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.13s, Blocked wall time: 0ns, Peak memory: 313.31MB, Memory allocations: 16519, Threads: 4, Splits: 10
```

Here is the printPlanWithStats of Project and Unnest before copyValuesAndNulls  optimization:
```
-- Unnest[c0, c1, c2] -> c0_e:BIGINT, c1_e:BIGINT, c2_e:BIGINT
   Output: 80357184 rows (1.98GB, 2688 batches), Cpu time: 2.73s, Blocked wall time: 0ns, Peak memory: 356.00KB, Memory allocations: 21504, Threads: 4
  -- Project[expressions: (c0:ARRAY<BIGINT>, array_constructor(ROW["l_suppkey"],ROW["l_suppkey"],ROW["l_suppkey"])), (c1:ARRAY<BIGINT>, array_constructor(ROW["l_orderkey"],ROW["l_orderkey"],null)), (c2:ARRAY<BIGINT>, array_constructor(ROW["l_partkey"],null,null))] -> c0:ARRAY<BIGINT>, c1:ARRAY<BIGINT>, c2:ARRAY<BIGINT>
     Output: 26785728 rows (2.72GB, 2688 batches), Cpu time: 21.04s, Blocked wall time: 0ns, Peak memory: 1.04MB, Memory allocations: 40332, Threads: 4
    -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
       Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.07s, Blocked wall time: 0ns, Peak memory: 308.52MB, Memory allocations: 16519, Threads: 4, Splits: 10

```

**Reply by mbasmanova** [reply]
@JkSelf Thank you for re-running the benchmark.

**Reply by mbasmanova** [reply]
FYI: #6567

**Comment by iwanttobepowerful** [comment]
> @mbasmanova Thanks for your help. This method facilitates a reduction in project time from 21s to 10s. However, the performance is still slower compared to the Expand operation.
> 
> Here is the printPlanWithStats of Expand:
> 
> ```
> -- Expand[[l_orderkey, l_partkey, l_suppkey, 0], [l_orderkey, null, l_suppkey, 1], [null, null, l_suppkey, 2]] -> l_orderkey:BIGINT, l_partkey:BIGINT, l_suppkey:BIGINT, group_id_0:BIGINT
>    Output: 80357184 rows (1.47GB, 8064 batches), Cpu time: 101.74ms, Blocked wall time: 0ns, Peak memory: 256B, Memory allocations: 8064, Threads: 4
>   -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
>      Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.32s, Blocked wall time: 0ns, Peak memory: 308.52MB, Memory allocations: 15388, Threads: 4, Splits: 10
> ```
> 
> Here is the printPlanWithStats of Project and Unnest after copyValuesAndNulls optimization:
> 
> ```
> -- Unnest[c0, c1, c2] -> c0_e:BIGINT, c1_e:BIGINT, c2_e:BIGINT
>    Output: 80357184 rows (1.98GB, 2688 batches), Cpu time: 2.82s, Blocked wall time: 0ns, Peak memory: 356.00KB, Memory allocations: 21504, Threads: 4
>   -- Project[expressions: (c0:ARRAY<BIGINT>, array_constructor(ROW["l_suppkey"],ROW["l_suppkey"],ROW["l_suppkey"])), (c1:ARRAY<BIGINT>, array_constructor(ROW["l_orderkey"],ROW["l_orderkey"],null)), (c2:ARRAY<BIGINT>, array_constructor(ROW["l_partkey"],null,null))] -> c0:ARRAY<BIGINT>, c1:ARRAY<BIGINT>, c2:ARRAY<BIGINT>
>      Output: 26785728 rows (2.72GB, 2688 batches), Cpu time: 10.29s, Blocked wall time: 0ns, Peak memory: 1.04MB, Memory allocations: 40332, Threads: 4
>     -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
>        Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.13s, Blocked wall time: 0ns, Peak memory: 313.31MB, Memory allocations: 16519, Threads: 4, Splits: 10
> ```
> 
> Here is the printPlanWithStats of Project and Unnest before copyValuesAndNulls optimization:
> 
> ```
> -- Unnest[c0, c1, c2] -> c0_e:BIGINT, c1_e:BIGINT, c2_e:BIGINT
>    Output: 80357184 rows (1.98GB, 2688 batches), Cpu time: 2.73s, Blocked wall time: 0ns, Peak memory: 356.00KB, Memory allocations: 21504, Threads: 4
>   -- Project[expressions: (c0:ARRAY<BIGINT>, array_constructor(ROW["l_suppkey"],ROW["l_suppkey"],ROW["l_suppkey"])), (c1:ARRAY<BIGINT>, array_constructor(ROW["l_orderkey"],ROW["l_orderkey"],null)), (c2:ARRAY<BIGINT>, array_constructor(ROW["l_partkey"],null,null))] -> c0:ARRAY<BIGINT>, c1:ARRAY<BIGINT>, c2:ARRAY<BIGINT>
>      Output: 26785728 rows (2.72GB, 2688 batches), Cpu time: 21.04s, Blocked wall time: 0ns, Peak memory: 1.04MB, Memory allocations: 40332, Threads: 4
>     -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
>        Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.07s, Blocked wall time: 0ns, Peak memory: 308.52MB, Memory allocations: 16519, Threads: 4, Splits: 10
> ```

@JkSelf Hi,  #6566 and #6568 . How much will performance improve after merging these two PRs?

**Reply by iwanttobepowerful** [reply]
21s to 10s?

**Reply by JkSelf** [reply]
@iwanttobepowerful Yes.
